File: impl/ocaml/src/at_variant.ml
(* Nested @variant expansion (SPEC §6.9). *)

open Ast
open Utils

(* Expands every nested @variant in place and returns the features. *)
let substitute (ast : nodes) (ds : Design_system.t) =
  let features = ref 0 in
  ignore
    (walk ast (fun n u ->
         match n with
         | At_rule at when at.name = "@variant" ->
             features := !features lor Features.variants;
             let groups = segment at.params ',' in
             let last = List.length groups - 1 in
             let result =
               List.concat
                 (List.mapi
                    (fun i group ->
                      let names = segment (String.trim group) ':' in
                      let nodes = if i < last then clone_nodes !(at.at_nodes) else !(at.at_nodes) in
                      let r = { selector = "&"; rule_nodes = ref nodes } in
                      List.iter
                        (fun name ->
                          let name = String.trim name in
                          if name = "" then
                            Twill_error.failf "Cannot use `@variant` with an empty variant name in `@variant %s`." at.params;
                          match Design_system.parse_variant ds name with
                          | None -> Twill_error.failf "Cannot use `@variant` with unknown variant: %s" name
                          | Some variant ->
                              if not (Variants.apply_variant (Rule r) variant ds.variants 0) then
                                Twill_error.failf "Cannot use `@variant` with variant: %s" name)
                        (List.rev names);
                      if r.selector = "&" then !(r.rule_nodes) else [ Rule r ])
                    groups)
             in
             replace_with u result;
             Continue
         | _ -> Continue));
  !features

File: impl/go/at_variant.go
package twill

import "strings"

// SubstituteAtVariant expands every nested @variant in place (SPEC §6.9).
func SubstituteAtVariant(ast *[]Node, ds *DesignSystem) (Features, error) {
	var features Features
	var failure error
	Walk(ast, func(n Node, u *WalkUtils) WalkAction {
		at, ok := n.(*AtRule)
		if !ok || at.Name != "@variant" {
			return Continue
		}
		features |= FeatureVariants
		groups := Segment(at.Params, ',')
		var result []Node
		for i, group := range groups {
			names := Segment(strings.TrimSpace(group), ':')
			children := at.Nodes
			if i < len(groups)-1 {
				children = CloneNodes(at.Nodes)
			}
			rule := StyleRule("&", children...)
			for j := len(names) - 1; j >= 0; j-- {
				name := strings.TrimSpace(names[j])
				if name == "" {
					failure = Errorf("Cannot use `@variant` with an empty variant name in `@variant %s`.", at.Params)
					return Stop
				}
				variant := ds.ParseVariant(name)
				if variant == nil {
					failure = Errorf("Cannot use `@variant` with unknown variant: %s", name)
					return Stop
				}
				if !ApplyVariant(rule, variant, ds.Variants, 0) {
					failure = Errorf("Cannot use `@variant` with variant: %s", name)
					return Stop
				}
			}
			if rule.Selector == "&" {
				result = append(result, rule.Nodes...)
			} else {
				result = append(result, rule)
			}
		}
		u.ReplaceWith(result...)
		return Continue
	})
	return features, failure
}
